File: facesdk_java_tests/src/test/java/PersonTests.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import com.epam.reportportal.junit5.ReportPortalExtension;
import com.regula.facesdk.webclient.ApiException;
import com.regula.facesdk.webclient.gen.model.*;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

@ExtendWith(ReportPortalExtension.class)
class PersonTests {

    private UUID groupId;
    private UUID personId;
    final String testGroupName = "test";
    private static final String nameA = "Person A";
    private static final String nameB = "Person B";
    Map<String, Object> baseMetadata = new HashMap<>() {{
        put("description", "This is a test group");
    }};
    private Person createdPerson;

    @BeforeEach
    void setUp() {
        try {
            // Create a test group
            GroupToCreate groupToCreate = new GroupToCreate();
            groupToCreate.setName(testGroupName);
            groupToCreate.setMetadata(baseMetadata);
            Group createdGroup = PathsConfig.faceSdk.groupApi.createGroup(groupToCreate);
            groupId = createdGroup.getId();
            assertNotNull(groupId, "No id field in the returned group");
            PersonFields personFields = new PersonFields();
            personFields.setName(nameA);
            personFields.setGroups(List.of(groupId));
            personFields.setMetadata(baseMetadata);

            createdPerson = PathsConfig.faceSdk.personApi.createPerson(personFields);
            personId = createdPerson.getId();
            assertNotNull(personId, "No id field in the returned person");
        } catch (ApiException e) {
            System.err.println("ApiException occurred during setup: " + e.getResponseBody());
            throw e;
        }
    }


    @AfterEach
    void tearDown() {
        // Delete the test person
        if (personId != null) {
            try {
                PathsConfig.faceSdk.personApi.deletePerson(personId);
            } catch (ApiException e) {
                System.out.println("Error during person deletion: " + e.getMessage());
            }
        }

        // Delete the test group
        if (groupId != null) {
            try {
                PathsConfig.faceSdk.groupApi.deleteGroup(groupId);
            } catch (ApiException e) {
                System.out.println("Error during group deletion: " + e.getMessage());
            }
        }
    }

    @Test
    void shouldCreatePerson() {
        try {
            // Assertions
            assertNotNull(createdPerson, "No response received");
            assertEquals(nameA, createdPerson.getName(), "The name of the person is incorrect");
            assertEquals(groupId, createdPerson.getGroups().get(0), "The groupId of the person is incorrect");
            assertEquals(baseMetadata.get("description"), createdPerson.getMetadata().get("description"), "The description of the group is incorrect");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }

    @Test
    void shouldGetAPerson() {
        try {
            // Get the test person
            Person retrievedPerson = PathsConfig.faceSdk.personApi.getPerson(personId);

            // Assertions
            assertNotNull(retrievedPerson, "No response received");
            assertEquals(personId, retrievedPerson.getId(), "Expected person ID to match");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }

    @Test
    void shouldUpdateAPerson() {
        try {
            // Update the test person
            String newDescription = "upd";
            Map<String, Object> updatedMetadata = new HashMap<>();
            updatedMetadata.put("description", newDescription);

            PersonToUpdateFields updatedPersonFields = new PersonToUpdateFields();
            updatedPersonFields.setName(nameB);
            updatedPersonFields.setMetadata(updatedMetadata);

            PathsConfig.faceSdk.personApi.updatePerson(personId, updatedPersonFields);

            // Get the updated person
            Person updatedPerson = PathsConfig.faceSdk.personApi.getPerson(personId);

            // Assertions
            assertNotNull(updatedPerson, "No response received");
            assertEquals(nameB, updatedPerson.getName(), "Person name not updated");
            assertEquals(newDescription, Objects.requireNonNull(updatedPerson.getMetadata()).get("description"), "Description not updated");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }

    @Test
    void shouldDeleteAPerson() {
        try {
            // Delete the test person
            PathsConfig.faceSdk.personApi.deletePerson(personId);

            // Get persons in the group and ensure the test person is deleted
            PersonsPage personsResponse = PathsConfig.faceSdk.groupApi.getAllPersonsByGroupId(1, 1, groupId);

            // Assertions
            assertNotNull(personsResponse, "No response received");
            assertTrue(Objects.requireNonNull(personsResponse.getItems()).isEmpty(), "Expected no persons in the group");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }

    @Test
    void testGetPersonImages() {
        try {
            byte[] imageContent = PathsConfig.readImageBytes(PathsConfig.FACE1_PATH);
            AddImageToPersonRequestImage image = new AddImageToPersonRequestImage();
            image.setContent(imageContent);
            AddImageToPersonRequest imageFields = new AddImageToPersonRequest();
            imageFields.setImage(image);
            PathsConfig.faceSdk.personApi.addImageToPerson(personId, imageFields);
            ImagePage addedImage = PathsConfig.faceSdk.personApi.getAllImagesByPersonId(1, 1, personId);
            assertNotNull(addedImage, "No response received");
            assertNotNull(addedImage.getItems().get(0).getId(), "Image ID should not be null");
            assertNotNull(addedImage.getItems().get(0).getPath(), "Image path should not be null");
        } catch (Exception e) {
            System.err.println("Exception occurred: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Test
    void shouldGetAllGroupsAssociatedWithAPerson() {
        try {
            // Create a second test group
            GroupToCreate secondGroupToCreate = new GroupToCreate();
            secondGroupToCreate.setName("secondTestGroup");
            secondGroupToCreate.setMetadata(baseMetadata);

            Group secondGroup = PathsConfig.faceSdk.groupApi.createGroup(secondGroupToCreate);
            UUID secondGroupId = secondGroup.getId();

            // Create a new person associated with both groups
            PersonFields newPersonFields = new PersonFields();
            newPersonFields.setName(nameA);
            newPersonFields.setGroups(Arrays.asList(groupId, secondGroupId)); // Assuming groupId is a UUID
            newPersonFields.setMetadata(baseMetadata);

            Person newPerson = PathsConfig.faceSdk.personApi.createPerson(newPersonFields);
            UUID newPersonId = newPerson.getId();
            GroupPage groupsResponse;
            groupsResponse = PathsConfig.faceSdk.personApi.getAllGroupsByPersonId(1, 2, newPersonId);

            // Assertions
            assertNotNull(groupsResponse, "No response received");
            assertNotNull(groupsResponse.getItems(), "No items in the response");
            assertEquals(2, groupsResponse.getItems().size(), "Expected 2 groups, but got a different number");
            assertTrue(groupsResponse.getItems().stream().anyMatch(group -> Objects.requireNonNull(group.getId()).equals(groupId)), "The first group not found in group IDs");
            assertTrue(groupsResponse.getItems().stream().anyMatch(group -> Objects.equals(group.getId(), secondGroupId)), "The second group not found in group IDs");

            // Clean up - delete the second group
            PathsConfig.faceSdk.groupApi.deleteGroup(secondGroupId);
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }

    @Test
    void shouldAddImageToPerson() {
        try {
            byte[] imageContent = PathsConfig.readImageBytes(PathsConfig.FACE1_PATH);
            AddImageToPersonRequestImage image = new AddImageToPersonRequestImage();
            image.setContent(imageContent);
            AddImageToPersonRequest imageFields = new AddImageToPersonRequest();
            imageFields.setImage(image);
            AddImageToPersonResponse addedImage = PathsConfig.faceSdk.personApi.addImageToPerson(personId, imageFields);
            // Assertions for the returned response
            assertNotNull(addedImage, "No response received");
            assertEquals("image/jpeg", addedImage.getContentType(), "Expected contentType to be 'image/jpeg'");
            assertNotNull(addedImage.getId(), "Image ID should not be null");

            // Get the updated person
            Person updatedPerson = PathsConfig.faceSdk.personApi.getPerson(personId);
            // Assertions
            assertNotNull(updatedPerson, "No response received");
            assertEquals(personId, updatedPerson.getId(), "Expected personId to match");
        } catch (Exception e) {
            System.err.println("Exception occurred: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }


    @Test
    void createPersonWithoutMetadata() {
        try {
            // Create a test group
            GroupToCreate groupToCreate = new GroupToCreate();
            groupToCreate.setName(testGroupName);
            Group createdGroup = PathsConfig.faceSdk.groupApi.createGroup(groupToCreate);
            UUID newGroupId = createdGroup.getId();
            assertNotNull(groupId, "No id field in the returned group");
            PersonFields personFields = new PersonFields();
            personFields.setName(nameA);
            personFields.setGroups(List.of(groupId)); // Assuming groupId is a UUID
            Person createdPerson = PathsConfig.faceSdk.personApi.createPerson(personFields);
            UUID newPersonId = createdPerson.getId();
            PathsConfig.faceSdk.personApi.deletePerson(newPersonId);
            PathsConfig.faceSdk.groupApi.deleteGroup(newGroupId);
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }
}

File: facesdk_cs_tests/facesdk_cs_tests/PersonTests.cs
namespace facesdk_cs_tests;

[TestFixture]
public class PersonTests
{
    private Guid? _groupId;
    private Guid? _personId;
    private const string TestGroupName = "test";
    private const string NameA = "Person A";
    private const string NameB = "Person B";
    private Person? _createdPerson;

    private readonly Dictionary<string, object> _baseMetadata = new Dictionary<string, object>
    {
        { "description", "This is a test group" }
    };

    [SetUp]
    public void SetUp()
    {
        try
        {
            // Create a test group
            var groupToCreate = new GroupToCreate
            {
                Name = TestGroupName,
                Metadata = _baseMetadata
            };
            var createdGroup = PathsConfig.FaceSdk.GroupApi.CreateGroup(groupToCreate);
            _groupId = createdGroup.Id;
            Assert.That(_groupId, Is.Not.Null, "No id field in the returned group");

            var personFields = new PersonFields
            {
                Name = NameA,
                Groups = new List<Guid> { (Guid)_groupId },
                Metadata = _baseMetadata
            };

            _createdPerson = PathsConfig.FaceSdk.PersonApi.CreatePerson(personFields);
            _personId = _createdPerson.Id;
            Assert.That(_personId, Is.Not.Null, "No id field in the returned person");
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred during setup: " + e.ErrorCode + e.Message);
            throw;
        }
    }

    [TearDown]
    public void TearDown()
    {
        // Delete the test group
        if (_groupId == Guid.Empty) return;
        {
            try
            {
                Debug.Assert(_groupId != null, nameof(_groupId) + " != null");
                PathsConfig.FaceSdk.GroupApi.DeleteGroup((Guid)_groupId);
            }
            catch (ApiException e)
            {
                Console.WriteLine("Error during group deletion: " + e.Message);
            }
        }

        // Delete the test person
        if (!_personId.HasValue) return;
        try
        {
            Debug.Assert(_personId != null, nameof(_personId) + " != null");
            PathsConfig.FaceSdk.PersonApi.DeletePerson((Guid)_personId);
        }
        catch (ApiException e)
        {
            Console.WriteLine("Error during person deletion: " + e.Message);
        }
    }

    [Test]
    public void ShouldCreateAPerson()
    {
        try
        {
            Assert.Multiple(() =>
            {
                Debug.Assert(_createdPerson != null, nameof(_createdPerson) + " != null");
                Assert.That(_createdPerson.Name, Is.EqualTo(NameA), "Expected person name to match");
                Assert.That(_createdPerson.Groups[0], Is.EqualTo(_groupId), "Expected person name to match");
                Assert.That(_createdPerson.Metadata?["description"], Is.EqualTo(_baseMetadata["description"]),
                    "The description of person is incorrect");
            });
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
    }

    [Test]
    public void ShouldGetAPerson()
    {
        try
        {
            // Get the test person
            Debug.Assert(_personId != null, nameof(_personId) + " != null");
            var retrievedPerson = PathsConfig.FaceSdk.PersonApi.GetPerson((Guid)_personId);

            // Assertions
            Assert.That(retrievedPerson, Is.Not.Null, "No response received");
            Assert.That(retrievedPerson.Id, Is.EqualTo(_personId), "Expected person ID to match");
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
    }

    [Test]
    public void ShouldUpdateAPerson()
    {
        try
        {
            // Update the test person
            const string newDescription = "upd";
            var updatedMetadata = new Dictionary<string, object>
            {
                { "description", newDescription }
            };

            var updatedPersonFields = new PersonFields
            {
                Name = NameB,
                Metadata = updatedMetadata
            };

            Debug.Assert(_personId != null, nameof(_personId) + " != null");
            PathsConfig.FaceSdk.PersonApi.UpdatePerson((Guid)_personId, updatedPersonFields);
            var updatedPerson = PathsConfig.FaceSdk.PersonApi.GetPerson((Guid)_personId);

            // Assertions
            Assert.That(updatedPerson.Name, Is.EqualTo(NameB), "Person name not updated");
            Assert.That(updatedPerson.Metadata?["description"], Is.EqualTo(newDescription), "Description not updated");
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
    }

    [Test]
    public void ShouldDeleteAPerson()
    {
        try
        {
            // Delete the test person
            Debug.Assert(_personId != null, nameof(_personId) + " != null");
            PathsConfig.FaceSdk.PersonApi.DeletePerson((Guid)_personId);

            // Get persons in the group and ensure the test person is deleted
            Debug.Assert(_groupId != null, nameof(_groupId) + " != null");
            var personsResponse = PathsConfig.FaceSdk.GroupApi.GetAllPersonsByGroupId(1, 1, (Guid)_groupId);

            // Assertions
            Assert.NotNull(personsResponse, "No response received");
            Assert.IsTrue(personsResponse.Items?.Count == 0, "Expected no persons in the group");
            _personId = null;
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
    }

    [Test]
    public void ShouldGetPersonImages()
    {
        try
        {
            var imageContent = PathsConfig.ReadImageBytes(PathsConfig.Face1Path);
            var image = new ImageFieldsImage
            {
                Content = imageContent
            };
            var imageFields = new ImageFields
            {
                Image = image
            };
            Debug.Assert(_personId != null, nameof(_personId) + " != null");
            PathsConfig.FaceSdk.PersonApi.AddImageToPerson((Guid)_personId, imageFields);
            var response =
                PathsConfig.FaceSdk.PersonApi.GetAllImagesByPersonId(page: 1, size: 1, personId: (Guid)_personId);

            // Assertions for the returned response
            Assert.That(response, Is.Not.Null, "No response received");
            Assert.That(response.Items[0].Id, Is.Not.Null, "id of photo not found");
            Assert.That(response.Items[0].Path, Is.Not.Null, "path of photo not found");
        }
        catch (Exception e)
        {
            Console.Error.WriteLine("Exception occurred: " + e.Message);
            throw;
        }
    }

    [Test]
    public void ShouldGetAllGroupsAssociatedWithAPerson()
    {
        try
        {
            // Create a second test group
            var secondGroupToCreate = new GroupToCreate
            {
                Name = "secondTestGroup",
                Metadata = _baseMetadata
            };
            var secondGroup = PathsConfig.FaceSdk.GroupApi.CreateGroup(secondGroupToCreate);
            var secondGroupId = secondGroup.Id;

            // Create a new person associated with both groups
            Debug.Assert(_groupId != null, nameof(_groupId) + " != null");
            var newPersonFields = new PersonFields
            {
                Name = NameA,
                Groups = new List<Guid> { (Guid)_groupId, secondGroupId },
                Metadata = _baseMetadata
            };
            var newPerson = PathsConfig.FaceSdk.PersonApi.CreatePerson(newPersonFields);
            var newPersonId = newPerson.Id;

            var groupsResponse = PathsConfig.FaceSdk.PersonApi.GetAllGroupsByPersonId(1, 2, newPersonId);

            // Assertions
            Assert.That(groupsResponse, Is.Not.Null, "No response received");
            Assert.That(groupsResponse.Items, Is.Not.Null, "No items in the response");
            Assert.That(groupsResponse.Items.Count, Is.EqualTo(2), "Expected 2 groups, but got a different number");
            Assert.That(groupsResponse.Items.Any(group => group.Id.Equals(_groupId)), Is.True,
                "The first group not found in group IDs");
            Assert.That(groupsResponse.Items.Any(group => group.Id.Equals(secondGroupId)), Is.True,
                "The second group not found in group IDs");

            // Clean up - delete the second group
            PathsConfig.FaceSdk.GroupApi.DeleteGroup(secondGroupId);
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
    }

    [Test]
    public void ShouldAddImageToPerson()
    {
        try
        {
            var imageContent = PathsConfig.ReadImageBytes(PathsConfig.Face1Path);
            var image = new ImageFieldsImage
            {
                Content = imageContent
            };
            var imageFields = new ImageFields
            {
                Image = image
            };
            Debug.Assert(_personId != null, nameof(_personId) + " != null");
            var addedImage = PathsConfig.FaceSdk.PersonApi.AddImageToPerson((Guid)_personId, imageFields);

            // Assertions for the returned response

            Assert.Multiple(() =>
            {
                Assert.That(addedImage, Is.Not.Null, "No response received");
                Assert.That(addedImage.ContentType, Is.EqualTo("image/jpeg"),
                    "Expected contentType to be 'image/jpeg'");
                Assert.That(addedImage.Id, Is.Not.Null, "Image ID should not be null");
                Assert.That(addedImage.Path, Is.Not.Null, "path of photo not found");
            });

            // Get the updated person
            var updatedPerson = PathsConfig.FaceSdk.PersonApi.GetPerson((Guid)_personId);

            // Assertions
            Assert.That(updatedPerson, Is.Not.Null, "No response received");
            Assert.That(updatedPerson.Id, Is.EqualTo(_personId), "Expected personId to match");
        }
        catch (Exception e)
        {
            Console.Error.WriteLine("Exception occurred: " + e.Message);
            throw;
        }
    }

    [Test]
    public void CreatePersonWithoutMetadata()
    {
        try
        {
            // Create a test group
            var groupToCreate = new GroupToCreate
            {
                Name = TestGroupName
            };
            var createdGroup = PathsConfig.FaceSdk.GroupApi.CreateGroup(groupToCreate);
            var newGroupId = createdGroup.Id;
            Assert.NotNull(newGroupId, "No id field in the returned group");

            var personFields = new PersonFields
            {
                Name = NameA,
                Groups = new List<Guid> { newGroupId }
            };
            var createdPersonWoMeta = PathsConfig.FaceSdk.PersonApi.CreatePerson(personFields);
            var newPersonId = createdPersonWoMeta.Id;

            PathsConfig.FaceSdk.PersonApi.DeletePerson(newPersonId);
            PathsConfig.FaceSdk.GroupApi.DeleteGroup(newGroupId);
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine("ApiException occurred: " + e.ErrorCode + e.Message);
            throw;
        }
    }
}